### ISSUE TITLE
Building Private Libraries/Executables with Lock Directory

### ISSUE THREAD

**Issue by rgrinberg** [ocaml/dune#8772]
Private libraries and executables are those that do not belong to any particular package. However, they still need to be built and may depend on public packages. There should be a way to specify which packages are necessary to build such libraries and executables.

**Comment by Alizter** [comment]
What if we allowed for `(package...)` in a deps like field?

**Comment by rgrinberg** [comment]
Where would the field go?

**Comment by rgrinberg** [comment]
Here's a simple suggestion:

1. Allow users to define a "private" package. That's a package for which we don't setup any install rules
2. Users can now attach additional dependencies/conflicts to this private package to build their private libraries/executables

**Comment by Alizter** [comment]
I was thinking of adding it to the library/executable stanza and make it mutually exclusive with package/public_name.

**Comment by rgrinberg** [comment]
See the discussion in https://github.com/ocaml/dune/pull/8356 for the problems with such an approach

**Comment by rgrinberg** [comment]
I don't think private packages are misleading and they have other uses.

Creating additional stanzas in the dune project file isn't great because it requires more work for conflicts and just about any other thing that can be attached to packages.

**Comment by Alizter** [comment]
I'm confused about the suggestion with private packages then. Where would you declare them?

**Comment by rgrinberg** [comment]
it would be the exact same thing as a normal package, except that it would have `(private)` set inside the dune-project file.

For such a package, there would be installation rules.

**Comment by Alizter** [comment]
How do we handle packages that depend on a private package? Will we need a similar setup like we have for private libraries and executables?

**Comment by rgrinberg** [comment]
For packages, we don't need separate private name and public name namespaces. We can just say that a public package cannot depend on a private package.

**Comment by Alizter** [comment]
Does this mean private packages can depend on other private packages? I feel like this is not a useful behaviour and I would rather have no package being able to depend on private packages.

**Comment by rgrinberg** [comment]
That's inconsistent with how we handle libraries and such dependencies are in fact useful. They can be used to group shared constraints between multiple private packages.

**Comment by Alizter** [comment]
How do we handle `(package ...)` dependencies in the dependency specification? This is really relying on the install layout so my instinct is to forbid this. Currently (with my PR) we get an error like this:
```
  Error: No rule found for alias .foo-files
  -> required by alias foo in dune:1
  [1]
```
since we are avoiding the install rules.

**Comment by rgrinberg** [comment]
For v1, you can forbid it. But it's possible to make it work. You just need to create a special "install" directory for every specification that uses a private package.

**Comment by Alizter** [comment]
It seems that Dune does not care about the `depends` field of `package` at all when building. I'm not really sure I can add any more validation here for private packages since there is no validation at all.

**Comment by rgrinberg** [comment]
Yes, for now we don't use the `depends` field for anything. It doesn't prevent you from adding your validation though. You could also add additional validation like cycle detection here.
